Send array of data to client, render entire table client side.

erwa opened the pull request:
Main changes:

* removed puzzle table HTML generation from server side. Instead, we just ship an array of data to the client, do all table rendering using JavaScript client-side. Data transferred reduced from 1.1 MB to 130 KB. Load time is now about 2 seconds on Heroku:
![transfer](https://user-images.githubusercontent.com/544734/71567823-f35efd00-2a8f-11ea-9323-be436cd762af.png)

* removed column-specific search boxes
* removed "Enter Guess" column, combined with "Answer" column, added new "Submit Answer" modal:
![answer-modal](https://user-images.githubusercontent.com/544734/71568062-2f935d00-2a92-11ea-8388-9b706daa80d7.png)

* made editing Status an on-demand modal:
![edit-status](https://user-images.githubusercontent.com/544734/71650893-5bf3e700-2ce7-11ea-91da-f77377f1b508.png)

* fixed bugs where modals weren't working when table was redrawn (e.g., on paging or filtering)
* increase default page size from 25 to 50
* reverted jquery datatables changes in https://github.com/cardinalitypuzzles/smallboard/pull/113/commits/bb6c55738556e5fdfe316fe1cc4e0cb34855e8e2 since they broke datatables searching

Screenshot:
![new-ui](https://user-images.githubusercontent.com/544734/71567811-cf032080-2a8f-11ea-9987-7b906fb75670.png)

For #114.

daiweili commented on puzzles/templates/modals/edit_status.html:
```diff
@@ -0,0 +1,41 @@
+            </div>
+            <form method="post" action="/puzzles/update_status/">
+                {% csrf_token %}
+                <div class="form-group">
```
add modal-body to get padding here

erwa replied:
Ah, thanks, the modal looked weird (no edge padding) and I was wondering why

daiweili commented on puzzles/templates/puzzles_table.html:
```diff
@@ -1,150 +1,174 @@
-        </tfoot>
-    </table>
-</div>
+<table class="table table-sm tree" id="puzzles" style="width:100%">
```
Wrap this in `<div class="px-3 pb-3"></div>` to get some extra padding as before. As is, the table has horizontal scroll.

erwa replied:
That looks much better. Thanks for noticing this.

daiweili commented on smallboard/static/jquery.dataTables.js:
```diff
@@ -4486,7 +4486,7 @@
 		var columns = settings.aoColumns;
```
Feel free to just delete this file and load from the CDN instead

erwa replied:
Good idea

daiweili commented on puzzles/templates/puzzles_table.html:
```diff
@@ -1,150 +1,174 @@
 <script type="text/javascript">
+var slack_base_url = '{{ slack_base_url }}';
+var data = [
+{% for p, tags, metas, active_users, status_form, puzzle_class in rows %}['{{p.pk}}', '{{p.name}}', '{{p.url}}', '{{p.is_meta}}', '{{active_users|join:", "}}', '{{p.answer}}', '{{p.status}}', '{{p.sheet}}', '{{p.channel}}', {{tags|safe}}, '', '{{puzzle_class}}', ],
```
No need to change for now, but in a future change, it'd make sense to expose an endpoint to just fetch the JSON data (rather than the page and having the data be delivered as part of the .html in a script tag)

erwa replied:
Yeah, I was thinking the same thing. Shouldn't be too hard to add such an endpoint, and then we should be able to refresh the table without an entire page reload. And we can set a pollInterval of, say, 5-10 seconds so the table refreshes automatically.

daiweili commented on puzzles/templates/puzzles_table.html:
```diff
@@ -1,150 +1,174 @@
+            {
+                'targets': tags,
+                'render': function(data, type, row, meta) {
+                    result = '';
```
nit: var result here and in all the other render functions?

erwa replied:
Updated. Good to avoid globals.

erwa commented:
Fixed the Answer Queue page. Removed the edit and delete puzzle functionality from the page.
![answer-queue](https://user-images.githubusercontent.com/544734/71650946-bdb45100-2ce7-11ea-8318-362d6cb834cd.png)

For now, I didn't optimize the answer queue page (all the HTML is still generated server-side). Will optimize it in a follow-up PR.

daiweili commented on smallboard/static/moment.min.js:
```diff
@@ -0,0 +1 @@
+!function(e,t){"object"==typeof exports&&"undefined"!=typeof module?module.exports=t():"function"==typeof define&&define.amd?define(t):e.moment=t()}(this,function(){"use strict";var e,i;function c(){return e.apply(null,arguments)}function o(e){return e instanceof Array||"[object Array]"===Object.prototype.toString.call(e)}function u(e){return null!=e&&"[object Object]"===Object.prototype.toString ...
```
Do you need to serve this? Does this CDN version not work: https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.24.0/moment.min.js

erwa replied:
Oh nice, using CDN is probably better. Will switch to this. Thanks.

daiweili commented on puzzles/templates/modals/edit_status.html:
```diff
@@ -1,41 +1,43 @@
+                <div class="modal-body">
+                    <div class="form-group">
+                        <select name="status" class="form-control form-control-sm">
+                            <option value="SOLVING" id='SOLVING'>SOLVING</option>
```
pre-existing, but why are these id's set to these all caps values? Are they needed?

erwa replied:
I think I added them to make it easier to get the DOM element using jQuery via `$('#<STATUS>')`. But looks like you can get by value using `$('#editstatus option[value="<STATUS>"])`. Removed the ids.
